backend/src/github-util.ts

```diff
@@ -0,0 +1,242 @@
+  // cannot get both with a single api call
+  return Promise.all([octokit.rest.pulls.get(pull_request), getReviewComments(pull_request)]).then(
+    ([pr, comments]) => ({
+      url: pr.data.html_url,
+      created_by: pr.data.user?.login || '',
+      comments
+    })
+  );
+};
+
+/** Retrieves information about opened PR `pull_request`. */
+const getOpenedPR = async (pull_request: PullRequest): Promise<OpenedPR> => {
+  const octokit = new Octokit();
+
+  return octokit.rest.pulls.get(pull_request).then((res) => {
+    // handle non 200 response?
+
+    const pr = res.data;
+    return {
+      url: pr.html_url,
+      created_by: pr.user?.login || '',
+      created_at: Date.parse(pr.created_at),
+      diff_size: pr.additions + pr.deletions
+    };
+  });
+};
+
+/** Creates 'valid' result if no errors are raised, 'invalid' result otherwise.  */
+const createValidationResult = async (validation_function): Promise<ValidationResult> => {
+  try {
+    await validation_function();
+  } catch (err) {
+    if (err instanceof Error) {
+      return {
+        status: 'invalid',
+        reason: err.message
+      };
+    }
+  }
+
+  // if no errors encountered, then valid submission
+  return { status: 'valid' };
+};
+
+/** Determines whether PR review is valid. */
+const validateReview = async (
+  portfolio: DevPortfolio,
+  submission: DevPortfolioSubmission,
+  review_url: string
+): Promise<ValidationResult> => {
+  const start = portfolio.earliestValidDate;
+  const end = portfolio.deadline;
+  const username = submission.member.github || ''; // must be github username
+
+  return createValidationResult(async () => {
+    // get review object
+    const review = await getReviewedPR(parseGithubUrl(review_url));
+
+    // cannot review own PR
+    if (review.created_by === username) {
+      throw new Error(`Cannot use PR ${review.url} opened by user for review requirement.`);
+    }
+
+    // comments made by user within the date range
+    const eligible_comments = filterComments(review.comments, username, start, end);
+
+    // placeholder logic: valid if at least 10 words total
+    const total_word_count = eligible_comments.reduce(
+      (count, comment) => count + comment.content.split(' ').length,
+      0
+    );
+    if (total_word_count < 10) {
+      throw new Error('Trivial review.');
+    }
+  });
+};
+
+/** Determines whether an open PR is valid. */
+const validateOpen = async (
+  portfolio: DevPortfolio,
+  submission: DevPortfolioSubmission,
+  open_url: string
+): Promise<ValidationResult> => {
+  const start = portfolio.earliestValidDate;
+  const end = portfolio.deadline;
+  const username = submission.member.github || ''; // must be github username
+
+  return createValidationResult(async () => {
+    // get open object
+    const open = await getOpenedPR(parseGithubUrl(open_url));
+
+    if (open.created_by !== username) {
+      throw new Error(`User ${username} did not open the pull request ${open.url}.`);
+    }
```

Comment:
I think as long as the member has commits in the PR then it's valid (i.e. they don't necessarily have to be the one opening the PR). I'm not 100% about this tho so let's leave this as-is for now.